File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/utils/RaycastUtils.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils;

import com.google.common.base.Predicates;
import net.minecraft.entity.Entity;
import net.minecraft.util.*;

import java.util.List;

public final class RaycastUtils extends MinecraftInstance {

    public static Entity raycastEntity(final double range, final IEntityFilter entityFilter) {
        return raycastEntity(range, RotationUtils.serverRotation.getYaw(), RotationUtils.serverRotation.getPitch(),
                entityFilter);
    }

    private static Entity raycastEntity(final double range, final float yaw, final float pitch, final IEntityFilter entityFilter) {
        final Entity renderViewEntity = mc.getRenderViewEntity();

        if(renderViewEntity != null && mc.theWorld != null) {
            double blockReachDistance = range;
            final Vec3 eyePosition = renderViewEntity.getPositionEyes(1F);

            final float yawCos = MathHelper.cos(-yaw * 0.017453292F - (float) Math.PI);
            final float yawSin = MathHelper.sin(-yaw * 0.017453292F - (float) Math.PI);
            final float pitchCos = -MathHelper.cos(-pitch * 0.017453292F);
            final float pitchSin = MathHelper.sin(-pitch * 0.017453292F);

            final Vec3 entityLook = new Vec3(yawSin * pitchCos, pitchSin, yawCos * pitchCos);
            final Vec3 vector = eyePosition.addVector(entityLook.xCoord * blockReachDistance, entityLook.yCoord * blockReachDistance, entityLook.zCoord * blockReachDistance);
            final List<Entity> entityList = mc.theWorld.getEntitiesInAABBexcluding(renderViewEntity, renderViewEntity.getEntityBoundingBox().addCoord(entityLook.xCoord * blockReachDistance, entityLook.yCoord * blockReachDistance, entityLook.zCoord * blockReachDistance).expand(1D, 1D, 1D), Predicates.and(EntitySelectors.NOT_SPECTATING, Entity :: canBeCollidedWith));

            Entity pointedEntity = null;

            for(final Entity entity : entityList) {
                if(!entityFilter.canRaycast(entity))
                    continue;

                final float collisionBorderSize = entity.getCollisionBorderSize();
                final AxisAlignedBB axisAlignedBB = entity.getEntityBoundingBox().expand(collisionBorderSize, collisionBorderSize, collisionBorderSize);
                final MovingObjectPosition movingObjectPosition = axisAlignedBB.calculateIntercept(eyePosition, vector);

                if(axisAlignedBB.isVecInside(eyePosition)) {
                    if(blockReachDistance >= 0.0D) {
                        pointedEntity = entity;
                        blockReachDistance = 0.0D;
                    }
                }else if(movingObjectPosition != null) {
                    final double eyeDistance = eyePosition.distanceTo(movingObjectPosition.hitVec);

                    if(eyeDistance < blockReachDistance || blockReachDistance == 0.0D) {
                        if(entity == renderViewEntity.ridingEntity && !renderViewEntity.canRiderInteract()) {
                            if(blockReachDistance == 0.0D)
                                pointedEntity = entity;
                        }else{
                            pointedEntity = entity;
                            blockReachDistance = eyeDistance;
                        }
                    }
                }
            }

            return pointedEntity;
        }

        return null;
    }

    public interface IEntityFilter {
        boolean canRaycast(final Entity entity);
    }
}

File: shared/main/java/net/ccbluex/liquidbounce/utils/RaycastUtils.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils

import net.ccbluex.liquidbounce.api.minecraft.client.entity.IEntity
import net.ccbluex.liquidbounce.api.minecraft.util.IAxisAlignedBB
import net.ccbluex.liquidbounce.api.minecraft.util.WMathHelper
import net.ccbluex.liquidbounce.api.minecraft.world.IWorld

object RaycastUtils : MinecraftInstance()
{

	@JvmStatic
	fun raycastEntity(theWorld: IWorld, thePlayer: IEntity, range: Double, entityFilter: (IEntity?) -> Boolean, aabbGetter: (IEntity) -> IAxisAlignedBB = IEntity::entityBoundingBox) = raycastEntity(theWorld, thePlayer, range, RotationUtils.serverRotation.yaw, RotationUtils.serverRotation.pitch, aabbGetter, entityFilter)

	@JvmStatic
	fun raycastEntity(theWorld: IWorld, thePlayer: IEntity, range: Double, yaw: Float, pitch: Float, aabbGetter: (IEntity) -> IAxisAlignedBB = IEntity::entityBoundingBox, entityFilter: (IEntity?) -> Boolean): IEntity?
	{
		val func = functions

		var reach = range
		val rayStartPos = thePlayer.getPositionEyes(1f)
		val ridingEntity = thePlayer.ridingEntity
		val canRiderInteract = thePlayer.canRiderInteract()

		val yawRadians = WMathHelper.toRadians(yaw)
		val pitchRadians = WMathHelper.toRadians(pitch)

		val yawCos = func.cos(-yawRadians - WMathHelper.PI)
		val yawSin = func.sin(-yawRadians - WMathHelper.PI)
		val pitchCos = -func.cos(-pitchRadians)
		val pitchSin = func.sin(-pitchRadians)

		val lookX = (yawSin * pitchCos).toDouble()
		val lookY = pitchSin.toDouble()
		val lookZ = (yawCos * pitchCos).toDouble()

		val rayEndPos = rayStartPos.addVector(lookX * reach, lookY * reach, lookZ * reach)

		val entityList = theWorld.getEntitiesInAABBexcluding(thePlayer, thePlayer.entityBoundingBox.addCoord(lookX * reach, lookY * reach, lookZ * reach).expand(1.0, 1.0, 1.0)) { it != null && (!classProvider.isEntityPlayer(it) || !it.asEntityPlayer().spectator) && it.canBeCollidedWith() }

		var pointedEntity: IEntity? = null

		entityList.filter(entityFilter::invoke).forEach { entity ->
			val collisionBorder = aabbGetter(entity)

			val rayIntercept = collisionBorder.calculateIntercept(rayStartPos, rayEndPos)

			if (collisionBorder.isVecInside(rayStartPos))
			{
				if (reach >= 0.0)
				{
					pointedEntity = entity
					reach = 0.0
				}
			}
			else if (rayIntercept != null)
			{
				val hitDistance = rayStartPos.distanceTo(rayIntercept.hitVec)

				if (hitDistance < reach || reach == 0.0) if (entity == ridingEntity && !canRiderInteract)
				{
					if (reach == 0.0) pointedEntity = entity
				}
				else
				{
					pointedEntity = entity
					reach = hitDistance
				}
			}
		}

		return pointedEntity
	}
}
